src/main/kotlin/org/jitsi/nlj/MediaSourceDesc.kt

```diff
@@ -0,0 +1,161 @@
+                layersById[encoding.encodingId(layer)] = layer
+                layersByIndex[layer.index] = layer
+            }
+        }
+    }
+
+    init { updateLayerCache() }
+
+    /**
+     * Gets the last "stable" bitrate (in bps) of the encoding of the specified
+     * index. The "stable" bitrate is measured on every new frame and with a
+     * 5000ms window.
+     *
+     * @return the last "stable" bitrate (bps) of the encoding at the specified
+     * index.
+     */
+    fun getBitrateBps(nowMs: Long, idx: Int): Long {
+        val layer = getRtpLayerByQualityIdx(idx) ?: return 0
+
+        // TODO: previous code returned a lower layer if this layer's bitrate was 0.
+        // Do we still need this?
+        return layer.getBitrateBps(nowMs)
+    }
+
+    fun hasRtpLayers(): Boolean =
+        synchronized(this) {
+            !layersByIndex.isEmpty()
+        }
+
+    /**
+     * Get an iterator over the source's RTP layers, in quality order.  Should be synchronized on [this].
+     */
+    val rtpLayers: Iterable<RtpLayerDesc>
+        get() = Iterable { Arrays.stream(rtpEncodings).flatMap { e: RtpEncodingDesc -> Arrays.stream(e.layers) }.iterator() }
+
+    fun numRtpLayers(): Int =
+        synchronized(this) {
+            layersByIndex.size
+        }
+
+    val primarySSRC: Long
+        get() = rtpEncodings[0].primarySSRC
+
+    fun getRtpLayerByQualityIdx(idx: Int): RtpLayerDesc? =
+        synchronized(this) {
+            layersByIndex[idx]
+        }
+
+    fun findRtpLayerDesc(videoRtpPacket: VideoRtpPacket): RtpLayerDesc? {
+        synchronized(this) {
+            if (ArrayUtils.isNullOrEmpty(rtpEncodings)) {
+                return null
+            }
+            val encodingId = getEncodingId(videoRtpPacket)
+            val desc = layersById[encodingId]
+            if (desc != null) {
+                return desc
+            }
+            /* ??? Does this part actually get used? */
+            for (encoding in rtpEncodings) {
+                if (encoding.matches(videoRtpPacket.ssrc)) {
+                    return encoding.findRtpLayerDesc(videoRtpPacket)
+                }
+            }
+            return null
+        }
+    }
+
+    private fun findRtpEncodingDesc(ssrc: Long): RtpEncodingDesc? =
+        synchronized(this) {
+            rtpEncodings.find { it.matches(ssrc) }
+        }
+
+    fun setEncodingLayers(layers: Array<RtpLayerDesc>, ssrc: Long) {
+        synchronized(this) {
+            val enc = findRtpEncodingDesc(ssrc) ?: return
+            enc.layers = layers
+            updateLayerCache()
+        }
+    }
+
+    override fun toString(): String {
+        val sb = StringBuilder()
+        sb.append("MediaSourceDesc ").append(hashCode()).append(" has encodings:\n")
+        sb.append(rtpEncodings.joinToString(separator = "\n  "))
+        return sb.toString()
+    }
+
+    /**
+     * FIXME: this should probably check whether the specified SSRC is part
```

Comment:
Does this need to be addressed before merging?

Reply:
This is an old FIXME copied from the Java -- so probably not?

Reply:
Since we're depacketizing RTX early on in the chain we it should be okay to only check the primary (media) ssrc here.